internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestEntries = 100
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	targetName string
+	remoteURL  string
+	entries    []lokiEntryWithService
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		httpClient: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) AddLog(entry servicelog.Entry) error {
+	c.entries = append(c.entries, lokiEntryWithService{
+		entry:   encodeEntry(entry),
+		service: entry.Service,
+	})
+	c.truncateBuffer()
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *Client) NumBuffered() int {
+	return len(c.entries)
+}
+
+func (c *Client) Flush(ctx context.Context) error {
+	if c.NumBuffered() == 0 {
+		return nil // no-op
+	}
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json; charset=utf-8")
+	httpReq.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	return c.handleServerResponse(resp)
+}
+
+// resetBuffer drops all buffered logs (in the case of a successful send, or an
+// unrecoverable error).
+func (c *Client) resetBuffer() {
+	c.entries = c.entries[:0]
+}
+
+// truncateBuffer truncates the buffer to maxRequestEntries, removing the
+// oldest logs first.
+func (c *Client) truncateBuffer() {
+	l := len(c.entries)
```

Comment:
IMO `num` or `numEntries` is clearer. Though at that point perhaps just `len(c.entries)` is clearest. The name `l` is a particularly confusing variable name as it looks so much like a 1 or I.